src/ripple/app/tx/impl/PayChan.cpp

```diff
@@ -395,90 +410,97 @@ PayChanClaim::doApply()
+            chanMembers[0][sfSequence],
+            chanMembers[1][sfSequence]};
 
-        if (ctx_.tx[~sfSignature])
+        for (auto const& txClaim : txClaims)
         {
-            PublicKey const pk ((*slep)[sfPublicKey]);
-            if (ctx_.tx[sfPublicKey] != pk)
+            auto const i =
+                txClaim[sfPublicKey] == chanMembers[0][sfPublicKey] ? 0 : 1;
+
+            if (i == 1 && txClaim[sfPublicKey] != chanMembers[1][sfPublicKey])
                 return temBAD_SIGNER;
-        }
 
-        if (reqBalance > chanFunds)
-            return tecUNFUNDED_PAYMENT;
+            if (txClaim[sfSequence] < chanMembers[i][sfSequence])
+                return tefPAST_SEQ;
 
-        if (reqBalance <= chanBalance)
-            // nothing requested
-            return tecUNFUNDED_PAYMENT;
+            if (txClaim[sfSequence] == chanMembers[i][sfSequence] &&
+                    txClaim[sfAmount] <= chanMembers[i][sfAmount])
+                return tefPAST_SEQ;
 
-        auto const sled = ctx_.view ().peek (keylet::account (dst));
-        if (!sled)
-            return terNO_ACCOUNT;
+            amounts[i] = txClaim[sfAmount];
+            sequences[i] = txClaim[sfSequence];
 
-        if (txAccount == src && ((*sled)[sfFlags] & lsfDisallowXRP))
-            return tecNO_TARGET;
+            if (ctx_.tx[sfAccount] == chanMembers[i].getAccountID (sfAccount))
+                submittedOwnClaim = true;
+        }
 
-        (*slep)[sfBalance] = ctx_.tx[sfBalance];
-        XRPAmount const reqDelta = reqBalance - chanBalance;
-        assert (reqDelta >= beast::zero);
-        (*sled)[sfBalance] = (*sled)[sfBalance] + reqDelta;
-        ctx_.view ().update (sled);
-        ctx_.view ().update (slep);
-    }
+        // Cannot submit single claim with higher sequence
+        if (sequences[0] != sequences[1])
+            return terPRE_SEQ;
 
-    if (ctx_.tx.getFlags () & tfRenew)
-    {
-        if (src != txAccount)
-            return tecNO_PERMISSION;
-        (*slep)[~sfExpiration] = boost::none;
-        ctx_.view ().update (slep);
+        if (chanMembers[0][sfBalance] < (amounts[0] - amounts[1]) ||
+            chanMembers[1][sfBalance] < (amounts[1] - amounts[0]))
+            return tecUNFUNDED_PAYMENT;
+
+        for (auto const& i : {0, 1})
+        {
+            chanMembers[i][sfAmount] = amounts[i];
+            chanMembers[i][sfSequence] = sequences[i];
+        }
+
+        slep->setFieldArray (sfChannelMembers, chanMembers);
     }
 
-    if (ctx_.tx.getFlags () & tfClose)
+    // Do not start the settle delay if third party submits claim.
+    // Do not reset settle delay if there is already an expiration,
+    // and a channel member submits only their own claim.
+    // This prevents a participant from perpetually keeping the channel open.
+    if ((! thirdPartySubmission || curExpiration) &&
+        (! curExpiration || ! submittedOwnClaim || txClaims.size() == 2))
```

Comment:
This is very hard to read. Using  de Morgan's laws can make it _slightly_ better.